Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp

```diff
@@ -701,24 +701,38 @@ void InspectorScopedShaderProgramHighlight::showHighlight()
 #endif
     }
 
-    saveBlendValue(GraphicsContextGL::BLEND_COLOR, m_savedBlend.color);
-    saveBlendValue(GraphicsContextGL::BLEND_EQUATION_RGB, m_savedBlend.equationRGB);
-    saveBlendValue(GraphicsContextGL::BLEND_EQUATION_ALPHA, m_savedBlend.equationAlpha);
-    saveBlendValue(GraphicsContextGL::BLEND_SRC_RGB, m_savedBlend.srcRGB);
-    saveBlendValue(GraphicsContextGL::BLEND_SRC_ALPHA, m_savedBlend.srcAlpha);
-    saveBlendValue(GraphicsContextGL::BLEND_DST_RGB, m_savedBlend.dstRGB);
-    saveBlendValue(GraphicsContextGL::BLEND_DST_ALPHA, m_savedBlend.dstAlpha);
-    saveBlendValue(GraphicsContextGL::BLEND, m_savedBlend.enabled);
+    auto& gl = *m_context.graphicsContextGL();
+
+    gl.getFloatv(GraphicsContextGL::BLEND_COLOR, m_blendColor);
+    m_blendState[0].equationRGB = gl.getInteger(GraphicsContextGL::BLEND_EQUATION_RGB);
+    m_blendState[0].equationAlpha = gl.getInteger(GraphicsContextGL::BLEND_EQUATION_ALPHA);
+    m_blendState[0].srcRGB = gl.getInteger(GraphicsContextGL::BLEND_SRC_RGB);
+    m_blendState[0].dstRGB = gl.getInteger(GraphicsContextGL::BLEND_DST_RGB);
+    m_blendState[0].srcAlpha = gl.getInteger(GraphicsContextGL::BLEND_SRC_ALPHA);
+    m_blendState[0].dstAlpha = gl.getInteger(GraphicsContextGL::BLEND_DST_ALPHA);
+    m_blendState[0].enabled = gl.isEnabled(GraphicsContextGL::BLEND);
+
+    if (m_context.m_oesDrawBuffersIndexed) {
+        for (unsigned i = 1; i < static_cast<unsigned>(m_context.getMaxDrawBuffers()); ++i) {
```

Comment:
So here we have logic:

1. save blend state
2. if draw buffers enabled, save draw buffers blend state
3. override blend state and draw buffers blend state
4. draw
5. restore saved blend state over blend state and draw buffers blend state
6. if draw buffers enabled, restore draw buffers blend state

Can we instead have:

1. if draw buffers enabled, save draw buffers blend state
2. else save blend state
3. draw
4. if draw buffers enabled, restore draw buffers blend state
5. else restore blend state

So in particular:
We do not want to save, override and restore 9 pieces of blend state: 1 blend state and 8 draw buffers blend states

We want to restore save, override and restore 1 blend state: either blend state or buffer 0 draw buffers blend state

Reply:
in addition, in case of draw buffers enabled, we want to save "is blend state 1-7 enabled", set them false and then restore them after draw..

Reply:
> So here we have logic:

`gl.getInteger(GraphicsContextGL::BLEND_EQUATION_RGB)` and friends return the state for draw buffer 0 when the `OES_draw_buffers_indexed` extension is enabled. The loop afterwards starts with 1. No value is queried or set more than once in `showHighlight`.